SUNDomEigEstimator Cleanup and remove extra vector memory allocation (#952)
LLNL/sundials: doc/shared/sundomeigest/SUNDomEigEst_API.rst

Reviewer: Note the newly added functions in the change log and recent changes files

      A :c:type:`SUNErrCode`.

.. c:function:: SUNErrCode SUNDomEigEstimator_SetPreRhsFn(SUNDomEigEstimator DEE, void* prerhs_fn_data, SUNPreRhsFn prerhs_fn)

   This *optional* function provides a :c:type:`SUNPreRhsFn` function for performing
   preprocessing of the right-hand side function, as well as a ``void*`` pointer to a data structure
   used by this routine, to the dominant eigenvalue estimator.

   **Arguments:**

      * *DEE* -- a SUNDomEigEstimator object.
      * *prerhs_fn_data* -- pointer to structure for ``prerhs_fn``.

>       * *prerhs_fn* -- function pointer to perform preprocessing of the right-hand side evaluations.  This is typically the same as the problem-defining function supplied to CVODE or ARKODE.

Reviewer 2: The `prerhs_fn` is called *before* the problem-defining function supplied to CVODE or ARKODE, so this comment is incorrect.  A more appropriate description of this function and its arguments would be:
```suggestion
   This *optional* function provides a :c:type:`SUNPreRhsFn` function for performing
   preprocessing of the right-hand side function, as well as a ``void*`` pointer to a data structure
   used by this routine, to the dominant eigenvalue estimator.  This *advanced* feature is similar to the ARKODE routine :c:func:`ARKPreRhsFn` -- we recommend that users read the warnings for that function before using this optional routine.

   **Arguments:**

      * *DEE* -- a SUNDomEigEstimator object.
      * *prerhs_fn_data* -- pointer to structure for ``prerhs_fn``.
      * *prerhs_fn* -- function pointer to perform preprocessing of the right-hand side evaluations.
```

      A :c:type:`SUNErrCode`.

.. c:function:: SUNErrCode SUNDomEigEstimator_SetPreRhsFn(SUNDomEigEstimator DEE, void* prerhs_fn_data, SUNPreRhsFn prerhs_fn)

   This *optional* function provides a :c:type:`SUNPreRhsFn` function for performing
   preprocessing of the right-hand side function, as well as a ``void*`` pointer to a data structure
   used by this routine, to the dominant eigenvalue estimator.  This *advanced* feature is similar to
   the ARKODE routine :c:func:`ARKPreRhsFn` -- we recommend that users read the warnings for 
   that function before using this optional routine.

   **Arguments:**

      * *DEE* -- a SUNDomEigEstimator object.
      * *prerhs_fn_data* -- pointer to structure for ``prerhs_fn``.
      * *prerhs_fn* -- function pointer to perform preprocessing of the right-hand side evaluations.

   **Return value:**

>       A :c:type:`SUNErrCode`.

Reviewer: ```suggestion
      A :c:type:`SUNErrCode`.

   .. versionadded:: x.y.z
```

      A :c:type:`SUNErrCode`.

.. c:function:: SUNErrCode SUNDomEigEstimator_SetRhsAtLinearizationPoint(SUNDomEigEstimator DEE, N_Vector Fyt)

   This *optional* function sets the value of the right-hand side function at the linearization point when using

>    :c:func:`SUNDomEigEstimator_SetRhs`.

Reviewer 2: ```suggestion
   :c:func:`SUNDomEigEstimator_SetRhs`.  When this is supplied, the *DEE* module 
   will use the value instead of re-evaluating the RHS at the linearization point.
```

      A :c:type:`SUNErrCode`.

.. c:function:: SUNErrCode SUNDomEigEstimator_SetRhsAtLinearizationPoint(SUNDomEigEstimator DEE, N_Vector Fyt)

   This *optional* function sets the value of the right-hand side function at the linearization point when using
   :c:func:`SUNDomEigEstimator_SetRhs`.  When this is supplied, the *DEE* module 

>    will use the value instead of re-evaluating the RHS at the linearization point.

Reviewer 2: ```suggestion
   will use the provided vector instead of re-evaluating the RHS at the linearization point.
```

      A :c:type:`SUNErrCode`.

.. c:function:: SUNErrCode SUNDomEigEstimator_SetRhsAtLinearizationPoint(SUNDomEigEstimator DEE, N_Vector Fyt)

   This *optional* function sets the value of the right-hand side function at the linearization point when using
   :c:func:`SUNDomEigEstimator_SetRhs`.  When this is supplied, the *DEE* module 
   will use the provided vector instead of re-evaluating the RHS at the linearization point.

   **Arguments:**

      * *DEE* -- a SUNDomEigEstimator object.
      * *Fyt* -- the value of the right-hand side function at the linearization point.

   **Return value:**

>       A :c:type:`SUNErrCode`.

Reviewer: ```suggestion
      A :c:type:`SUNErrCode`.

   .. versionadded:: x.y.z
```

   to do its job. The time parameter :math:`t` and the vector :math:`y` should be left 
   unchanged.

.. c:type:: int (*SUNPreRhsFn)(sunrealtype t, N_Vector y, void* prerhs_fn_data)

   Used to perform any preprocessing of the right-hand side function before
   performing a discrete Jacobian-vector product using quotient approximations of the Jacobian.
   The parameter *prerhs_fn_data* is a pointer to any information about RHS which the function needs in order 
   to do its job. The time parameter :math:`t` and the vector :math:`y` should be left 

>    unchanged.

Reviewer 2: ```suggestion
   Used to perform any preprocessing of the right-hand side function before
   performing a Jacobian-vector product using difference quotient approximations.
   The parameter *prerhs_fn_data* is a pointer to any problem-defining information that this function needs in order 
   to do its job. Both the time parameter :math:`t` and the vector :math:`y` inputs should be left 
   unchanged.
```

   to do its job. The time parameter :math:`t` and the vector :math:`y` should be left 
   unchanged.

.. c:type:: int (*SUNPreRhsFn)(sunrealtype t, N_Vector y, void* prerhs_fn_data)

   Used to perform any preprocessing of the right-hand side function before
   performing a Jacobian-vector product using difference quotient approximations.
   The parameter *prerhs_fn_data* is a pointer to any problem-defining information
   that this function needs in order to do its job. Both the time parameter

>    :math:`t` and the vector :math:`y` inputs should be left unchanged.

Reviewer: ```suggestion
   :math:`t` and the vector :math:`y` inputs should be left unchanged.
   
   .. versionadded:: x.y.z
```
